src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+
+from .functions import FloatType
+from .operators import IterativeOperator, TimeDependentOperator
+
+__all__ = [
+    "SecondaryOperator",
+    "SecondaryExpression",
+]
+
+
+class SecondaryOperator(
+    TimeDependentOperator,
+    IterativeOperator,
+):
+    """Operator representing a :class:`SecondaryExpression` in AD operator form on
+    specified subdomains or interfaces, at a time or iterate index.
+
+    Not meant to be instantiated directly, only by calling the factory class
+    :class:`SecondaryExpression`.
+
+    The secondary operator is essentially like a
+    :class:`~porepy.numerics.ad.operator_functions.Function`. Upon evaluation it fetches
+    the stored values and the derivative values.
+    The derivative values are inserted into the Jacobians of the first-order
+    dependencies (identity blocks).
+
+    Parameters:
+        name: Name of the called :class:`SecondaryExpression`.
+        domains: Arguments to its call.
+        children: The first-order dependencies of the called
+            :class:`SecondaryExpression` in AD form (defined on the same ``domains``).
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        domains: Sequence[pp.Grid] | Sequence[pp.MortarGrid],
+        children: Sequence[pp.ad.Variable],
+    ) -> None:
+
+        super().__init__(
+            name=name,
+            domains=domains,
+            operation=pp.ad.Operator.Operations.evaluate,
+            children=children,
+        )
+
+        self.fetch_data: Callable[[SecondaryOperator, pp.GridLike, bool], np.ndarray]
+        """A function returning the stored data for a secondary operator on a grid,
+        which is time step and iterate index dependent.
+
+        This function is assigned by the factory class :class:`SecondaryExpression`
+        which has a reference to the md-grid and must not be touched by the user.
+
+        """
+
+    def __repr__(self) -> str:
+        """String representation giving information on name, time and iterate index, as
+        well as domains and dependencies."""
+
+        msg = (
+            f"Secondary operator {self.name}.\n"
+            + f"\nDefined on {len(self._domains)} {self._domain_type}.\n"
+            + f"Dependent on {len(self.children)} independent expressions.\n"
+        )
+
+        if self.is_previous_time:
+            msg += f"Evaluated at the previous time step {self.time_step_index}.\n"
+        elif self.is_previous_iterate:
+            msg += f"Evaluated at the previous iterate {self.iterate_index}.\n"
+
+        return msg
+
+    def __call__(self, *args: Operator) -> Operator:
+        """By calling this operator with new children, the user can change the
+        structure of the Jacobian.
+
+        The dependencies/children are replaced and hence the non-trivial entries in
+        the Jacobian in the md-setting.
+
+        Note:
+            Call this operator only if you know what you are doing.
```

Comment:
I have 2 functionalities in mind:

The user can restrict the represented expression to a subset of grids, without going through the machinery of the factory class, and without messing with time and iterate index.
The user can dynamically change the dependencies (and hence the structure of the Jacobian), but this is probably insane and I can't think of a use case.

Since it also is a kind of artifact from the original intention to inherit from abstract function, I think the call-overload can be removed.

Reply:
If you think we can remove this, I suggest we do so.

FYI: Restricting operators to a subset of grids may be part of a future enhancement of the Ad framework.

Reply:
Call-overload removed.